packages/browser/src/index.test.ts

```diff
@@ -1,17 +1,51 @@
-import { createRequester } from '@logto/js';
+import { generateCodeChallenge, parseUriParameters, QueryKey } from '@logto/js';
 
-import LogtoClient from './index';
+import LogtoClient, { getLogtoKey, LogtoSignInSessionItem } from './index';
+
+const clientId = 'client_id_value';
+const endpoint = 'https://logto.dev';
+const authorizationEndpoint = `${endpoint}/oidc/auth`;
+const requester = jest.fn();
+
+jest.mock('@logto/js', () => ({
+  ...jest.requireActual('@logto/js'),
+  fetchOidcConfig: async () => ({
+    authorizationEndpoint,
+  }),
+}));
 
 describe('LogtoClient', () => {
   test('constructor', () => {
-    expect(
-      () =>
-        new LogtoClient({
-          endpoint: 'https://logto.dev',
-          clientId: 'client_id_value',
-          usingPersistStorage: false,
-          requester: createRequester(),
-        })
-    ).not.toThrow();
+    expect(() => new LogtoClient({ endpoint, clientId, requester })).not.toThrow();
+  });
+
+  describe('signIn', () => {
+    test('sessionStorage and signInUri should contain correct properties', async () => {
+      const logtoClient = new LogtoClient({ endpoint, clientId, requester });
+      const originalRedirectUri = 'http://localhost:3000/callback';
+      await logtoClient.signIn(originalRedirectUri);
+
+      const item = sessionStorage.getItem(getLogtoKey(clientId));
+      expect(item).not.toBeNull();
+
+      if (item) {
+        const logtoSignInSessionItem = JSON.parse(item) as LogtoSignInSessionItem;
+        const { state, codeVerifier, redirectUri } = logtoSignInSessionItem;
+        const codeChallenge = await generateCodeChallenge(codeVerifier);
+
+        expect(redirectUri).toEqual(originalRedirectUri);
+        expect(state.length).toBeGreaterThan(1);
+        expect(codeVerifier.length).toBeGreaterThan(1);
+
+        const signInUri = window.location.toString();
+        expect(signInUri.startsWith(authorizationEndpoint));
+
+        const urlSearchParameters = parseUriParameters(signInUri);
+        expect(urlSearchParameters.get(QueryKey.ClientId)).toEqual(clientId);
+        expect(urlSearchParameters.get(QueryKey.RedirectUri)).toEqual(originalRedirectUri);
+        expect(urlSearchParameters.get(QueryKey.CodeChallenge)).toEqual(codeChallenge);
+        expect(urlSearchParameters.get(QueryKey.State)).toEqual(state);
```

Comment:
better split into two different test cases. 
1. generate signInUrl
2. session storage

Reply:
The source code no longer exists and I resolve this commement.